### PR TITLE
Inconsistent datetime formats in Course Catalog API

### PR DESCRIPTION
**Pull request by ahsan-ul-haq**
ECOM-4354

@schenedx please review

### CODE REVIEW

**Review by clintonb — CHANGES_REQUESTED** [review]
We do NOT want fractional seconds in the API responses. Also, this should be handle via settings: http://www.django-rest-framework.org/api-guide/settings/#date-and-time-formatting

**Comment by clintonb** [comment]
See https://github.com/tomchristie/django-rest-framework/blob/a68b37d8bc432fae37ef5880aec500002b59f565/rest_framework/utils/humanize_datetime.py#L10.

We want to use that format without the optional milliseconds. After discussing with @schenedx, always show milliseconds: `'YYYY-MM-DDThh:mm:ss.uuuuuu[+HH:MM|-HH:MM|Z]'`

**Comment by ahsan-ul-haq** [comment]
@clintonb I think with %f it will always show milliseconds. Do I need to do something else?

**Comment by clintonb** [comment]
Use the setting intead of modifying each serializer field.

**Comment by ahsan-ul-haq** [comment]
@clintonb please review

File: course_discovery/apps/api/serializers.py [hunk 1]
```diff
@@ -309,7 +309,7 @@ class SeatSerializer(serializers.ModelSerializer):
         max_digits=Seat.PRICE_FIELD_CONFIG['max_digits']
     )
     currency = serializers.SlugRelatedField(read_only=True, slug_field='code')
-    upgrade_deadline = serializers.DateTimeField()
+    upgrade_deadline = serializers.DateTimeField(format=settings.API_DATETIME_FORMAT)
```
Comment (clintonb) [inline course_discovery/apps/api/serializers.py]: You should not need to change any of these serializers.

File: course_discovery/apps/api/v1/views/course_runs.py [hunk 2]
```diff
@@ -87,7 +87,7 @@ def list(self, request, *args, **kwargs):
               paramType: query
               multiple: false
             - name: hidden
-              description: Filter based on wether the course run is hidden from search.
+              description: Filter based on weather the course run is hidden from search.
```
Comment (clintonb) [inline course_discovery/apps/api/v1/views/course_runs.py]: whether

File: course_discovery/settings/base.py [hunk 3]
```diff
@@ -479,3 +479,7 @@
         'debug_toolbar.panels.redirects.RedirectsPanel',
         'elastic_panel.panel.ElasticDebugPanel'
     ]
+
+
+# API Datetime Format
+API_DATETIME_FORMAT = '%Y-%m-%dT%H:%M:%S.%fZ'
```
Comment (clintonb) [inline course_discovery/settings/base.py]: http://www.django-rest-framework.org/api-guide/settings/#date-and-time-formatting

```python
REST_FRAMEWORK = {
  # NOTE: This is the same format as DRF's iso-8601, however the milliseconds are always rendered.
  'DATETIME_FORMAT': 'YYYY-MM-DDThh:mm:ss.uuuuuu[+HH:MM|-HH:MM|Z]',
}
```

**Review by clintonb — CHANGES_REQUESTED** [review]

**Comment by ahsan-ul-haq** [comment]
@clintonb made the change

File: course_discovery/settings/base.py [hunk 4]
```diff
@@ -333,6 +333,8 @@
     'DEFAULT_THROTTLE_RATES': {
         'user': '100/hour',
     },
+    # NOTE: This is the same format as DRF's iso-8601, however the milliseconds are always rendered.
+    'DATETIME_FORMAT': '%Y-%m-%dT%H:%M:%S.%fZ',
```
Comment (clintonb) [inline course_discovery/settings/base.py]: Please use this exact string: `'YYYY-MM-DDThh:mm:ss.uuuuuu[+HH:MM|-HH:MM|Z]'` to avoid issues with time zones.

Comment (ahsan-ul-haq) [reply course_discovery/settings/base.py]: @clintonb Doing that causes the dates time to appear like following
screen shot 2017-02-15 at 7 49 14 pm

**Comment by ahsan-ul-haq** [comment]
I have made change for time zone as well

File: course_discovery/settings/base.py (hunk 4 above)
Comment (schenedx) [reply course_discovery/settings/base.py]: @clintonb I read the DRF documentation again. They require python strftime format string. See http://www.django-rest-framework.org/api-guide/settings/#datetime_format

So `YYYY-MM-DDThh:mm:ss.uuuuuu[+HH:MM|-HH:MM|Z]` is not a python strftime format. It's just a format they use on datetime string to replace `iso-8601`.

**Review by schenedx — APPROVED** [review]
👍

**Comment by schenedx** [comment]
@ahsan-ul-haq You got 8 unit test failures, could you please investigate?

**Pull request closed without merging**